geopandas/geodataframe.py

```diff
@@ -211,6 +211,8 @@ def __init__(
                 or geom_dtype == "object"
                 # special case for geometry = [], has float dtype
                 or (len(self) == 0 and geom_dtype == "float")
+                # special case for geometry = [np.nan]
+                or ((not self.empty) and pd.isna(self["geometry"]).all())
```

Comment:
```suggestion
                or ((not self.empty) and self["geometry"].isna().all())
```

(will do the same, but I find it slightly more readable)